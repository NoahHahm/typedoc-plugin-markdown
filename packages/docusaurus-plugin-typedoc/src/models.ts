import { TypedocPluginMarkdownOptions } from 'typedoc-plugin-markdown';

export interface PluginOptions extends TypedocPluginMarkdownOptions {
  id: string;
  docsRoot: string;
  sidebar: SidebarOptions;
}

export interface SidebarOptions {
  categoryLabel: string;
  position: number | null;
  autoConfiguration: boolean;
  indexLabel: string;
}

export interface SidebarCategory {
  type: string;
  label: string;
  items: SidebarItem[];
}

export interface CategoryYamlOptions {
  path: string;
  label: string;
  position: number | null;
  collapsed: boolean;
}

export type SidebarItem = SidebarCategory | string;
