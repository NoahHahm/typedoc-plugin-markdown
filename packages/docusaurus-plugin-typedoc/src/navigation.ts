import { ReflectionKind } from 'typedoc';

export const CATEGORY_POSITION = {
  [ReflectionKind.Module]: 1,
  [ReflectionKind.Namespace]: 1,
  [ReflectionKind.Enum]: 2,
  [ReflectionKind.Class]: 3,
  [ReflectionKind.Interface]: 4,
  [ReflectionKind.TypeAlias]: 5,
  [ReflectionKind.Variable]: 6,
  [ReflectionKind.Function]: 7,
  [ReflectionKind.ObjectLiteral]: 8,
};
