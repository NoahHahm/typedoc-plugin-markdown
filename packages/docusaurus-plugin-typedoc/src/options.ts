import { PluginOptions } from './types';

const DEFAULT_PLUGIN_OPTIONS: PluginOptions = {
  id: 'default',
  docsRoot: 'docs',
  out: 'api',
  cleanOutputDir: true,
  sidebar: {
    fullNames: false,
    categoryLabel: 'API',
    collapsed: true,
    indexLabel: 'Overview',
    position: null,
    autoConfiguration: true,
  },
  hideInPageTOC: true,
  hideBreadcrumbs: true,
  hidePageTitle: false,
  entryDocument: 'index.md',
  plugin: ['none'],
  watch: false,
  theme: 'docusaurus',
  enableFrontmatter: true,
  enableEmojis: true,
};

export const getPluginOptions = (
  opts: Partial<PluginOptions>,
): PluginOptions => {
  const options = {
    ...DEFAULT_PLUGIN_OPTIONS,
    ...opts,
    sidebar: {
      ...DEFAULT_PLUGIN_OPTIONS.sidebar,
      ...opts.sidebar,
    },
  };
  return options;
};
