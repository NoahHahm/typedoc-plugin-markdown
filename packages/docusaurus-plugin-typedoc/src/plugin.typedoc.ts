import * as fs from 'fs';
import {
  Application,
  DeclarationReflection,
  MixedDeclarationOption,
  PageEvent,
  ParameterType,
  RendererEvent,
} from 'typedoc';
import { CategoryYamlOptions, SidebarOptions } from './models';

export function load(app: Application) {
  app.options.addDeclaration({
    name: 'sidebar',
    type: ParameterType.Mixed,
  } as MixedDeclarationOption);

  app.renderer.on(PageEvent.END, (page: PageEvent<DeclarationReflection>) => {
    if (page.contents) {
      page.contents = page.contents.replace(/\\</g, '<');
    }
  });

  app.renderer.postRenderAsyncJobs.push(async (renderer: RendererEvent) => {
    const sidebar = app.options.getValue('sidebar') as SidebarOptions;
    if (sidebar.autoConfiguration) {
      writeCategoryYaml({
        path: renderer.outputDirectory,
        label: sidebar.categoryLabel,
        position: sidebar.position,
        collapsed: sidebar.collapsed,
      });
    }
  });
}

function writeCategoryYaml(categoryYamlOptions: CategoryYamlOptions) {
  const yaml: string[] = [`label: "${categoryYamlOptions.label}"`];
  if (categoryYamlOptions.position !== null) {
    yaml.push(`position: ${categoryYamlOptions.position}`);
  }
  if (!categoryYamlOptions.collapsed) {
    yaml.push(`collapsed: false`);
  }
  if (fs.existsSync(categoryYamlOptions.path)) {
    fs.writeFileSync(
      categoryYamlOptions.path + '/_category_.yml',
      yaml.join('\n'),
    );
  }
}
