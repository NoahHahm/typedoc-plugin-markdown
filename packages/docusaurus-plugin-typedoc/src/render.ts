import {
  Application,
  MixedDeclarationOption,
  ParameterType,
  StringDeclarationOption,
  TSConfigReader,
  TypeDocReader,
} from 'typedoc';

export function addTypedocReaders(app: Application) {
  app.options.addReader(new TypeDocReader());
  app.options.addReader(new TSConfigReader());
}

export function addTypedocDeclarations(app: Application) {
  app.options.addDeclaration({
    name: 'id',
  } as StringDeclarationOption);

  app.options.addDeclaration({
    name: 'docsRoot',
  } as StringDeclarationOption);

  app.options.addDeclaration({
    name: 'globalsTitle',
  } as StringDeclarationOption);

  app.options.addDeclaration({
    name: 'sidebar',
    type: ParameterType.Mixed,
  } as MixedDeclarationOption);
}
