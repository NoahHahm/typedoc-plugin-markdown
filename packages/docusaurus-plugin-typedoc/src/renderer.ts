import {
  Application,
  DeclarationReflection,
  PageEvent,
  ReflectionKind,
} from 'typedoc';
import { FrontmatterEvent } from 'typedoc-plugin-frontmatter';
import { MarkdownTheme } from 'typedoc-plugin-markdown';
import { SidebarOptions } from './models';

export function loadRenderer(app: Application) {
  app.renderer.on(PageEvent.END, (page: PageEvent<DeclarationReflection>) => {
    if (page.contents) {
      page.contents = page.contents.replace(/\\</g, '<');
    }
  });

  app.renderer.on(
    FrontmatterEvent.PREPARE_FRONTMATTER,
    (event: FrontmatterEvent) => {
      loadFrontmatter(app, event);
    },
  );
}

function loadFrontmatter(app: Application, event: FrontmatterEvent) {
  const renderContext = (
    app.renderer.theme as MarkdownTheme
  ).getRenderContext();

  const hasReadme = !app.options.getValue('readme').endsWith('none');
  const sidebar = app.options.getValue('sidebar') as SidebarOptions;
  const isModulesPage = event.page?.url.endsWith(renderContext.modulesFile);
  const isPackagesPage = event.page?.url.endsWith(renderContext.packagesFile);
  const isReadmePage = event.page?.url.endsWith(
    app.options.getValue('entryDocument') as string,
  );
  const isPackageModule =
    event.page.model.kindOf(ReflectionKind.Module) &&
    !Boolean((event.page.model?.parent as any).groups);
  const model = event.page?.model;

  let pluginFrontmatter: Record<string, any> = {};

  if (sidebar.autoConfiguration) {
    // Add sidebar labels to all pages
    if (!isModulesPage && !isReadmePage) {
      const sidebarLabel =
        !model.kindOf(ReflectionKind.Module) &&
        !app.options.getValue('hideKindTag') &&
        app.options.getValue('excludeGroups')
          ? `(${Array.from(ReflectionKind.singularString(model.kind))[0]}) ${
              model?.name
            }`
          : model?.name;
      pluginFrontmatter = {
        ...pluginFrontmatter,
        sidebar_label: sidebarLabel,
      };
    }
    // Entry url
    if (
      event.page?.url === app.options.getValue('entryDocument') ||
      (!hasReadme && isModulesPage)
    ) {
      if (sidebar.categoryLabel) {
        pluginFrontmatter = {
          ...pluginFrontmatter,
          sidebar_label: sidebar.categoryLabel,
        };
      }

      if (sidebar.position !== null && !Number.isNaN(sidebar.position)) {
        pluginFrontmatter = {
          ...pluginFrontmatter,
          sidebar_position: sidebar.position,
        };
      }
    }

    if (isModulesPage) {
      pluginFrontmatter = {
        ...pluginFrontmatter,
        sidebar_label: sidebar.indexLabel,
      };
    }

    if (isPackageModule && isReadmePage) {
      pluginFrontmatter = {
        ...pluginFrontmatter,
        sidebar_label: event.page.model.name,
      };
    }

    if (isPackagesPage) {
      pluginFrontmatter = {
        ...pluginFrontmatter,
        sidebar_label: 'Packages',
      };
    }

    if (isPackageModule && isModulesPage && Boolean(event.page.model?.readme)) {
      pluginFrontmatter = {
        ...pluginFrontmatter,
        sidebar_label: sidebar.indexLabel,
      };
    }

    if (
      isPackageModule &&
      isModulesPage &&
      !Boolean(event.page.model?.readme)
    ) {
      pluginFrontmatter = {
        ...pluginFrontmatter,
        sidebar_label: event.page.model.name,
      };
    }
  }

  event.frontmatter = {
    ...pluginFrontmatter,
    ...event.frontmatter,
  };
}
