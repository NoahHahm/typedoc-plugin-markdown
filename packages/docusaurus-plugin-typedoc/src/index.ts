import * as path from 'path';
import { Application, TypeDocOptions } from 'typedoc';
import { PluginOptions } from './models';
import { getPluginOptions, loadOptions } from './options';
import { loadRenderer } from './renderer';

// store list of plugin ids when running multiple instances
const apps: string[] = [];

export default function pluginDocusaurus(
  context: any,
  opts: Partial<PluginOptions>,
) {
  return {
    name: 'docusaurus-plugin-typedoc',
    async loadContent() {
      if (opts.id && !apps.includes(opts.id)) {
        apps.push(opts.id);
        generateTypedoc(context, opts);
      }
    },
    extendCli(cli) {
      cli
        .command('generate-typedoc')
        .description(
          '(docusaurus-plugin-typedoc) Generate TypeDoc docs independently of the Docusaurus build process.',
        )
        .action(async () => {
          context.siteConfig?.plugins.forEach((pluginConfig) => {
            // Check PluginConfig is typed to [string, PluginOptions]
            if (pluginConfig && typeof pluginConfig[1] === 'object') {
              generateTypedoc(context, pluginConfig[1]);
            }
          });
        });
    },
  };
}

/**
 * Initiates a new typedoc Application bootstraped with plugin options
 */
async function generateTypedoc(context: any, opts: Partial<PluginOptions>) {
  const { siteDir } = context;

  const options = getPluginOptions(opts);

  const { id, docsRoot, ...optionsPassedToTypeDoc } = options;

  const outputDir = path.resolve(siteDir, options.docsRoot, options.out);

  const app = new Application();

  loadOptions(app);

  loadRenderer(app);

  await app.bootstrapWithPlugins(
    optionsPassedToTypeDoc as unknown as Partial<TypeDocOptions>,
  );

  const project = app.convert();

  // if project is undefined typedoc has a problem - error logging will be supplied by typedoc.
  if (!project) {
    return;
  }

  if (options.watch) {
    app.convertAndWatch(async (project) => {
      await app.generateDocs(project, outputDir);
    });
  } else {
    await app.generateDocs(project, outputDir);
  }
}
