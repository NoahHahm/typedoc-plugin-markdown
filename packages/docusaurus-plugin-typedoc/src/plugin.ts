import * as fs from 'fs';
import * as path from 'path';
import { Application, TypeDocOptions } from 'typedoc';
import { load as loadTypedocPluginFrontmatter } from 'typedoc-plugin-frontmatter';
import { load as loadTypedocPluginMarkdown } from 'typedoc-plugin-markdown';
import { PluginOptions } from '.';
import { getPluginOptions, loadOptions } from './options';
import { loadRenderer } from './renderer';

// store list of plugin ids when running multiple instances
const apps: string[] = [];

export default function pluginDocusaurus(
  context: any,
  opts: Partial<PluginOptions>,
) {
  return {
    name: 'docusaurus-plugin-typedoc',
    async loadContent() {
      if (opts.id && !apps.includes(opts.id)) {
        apps.push(opts.id);
        await generateTypedoc(context, opts);
      }
    },
    extendCli(cli) {
      cli
        .command('generate-typedoc')
        .description(
          '(docusaurus-plugin-typedoc) Generate TypeDoc docs independently of the Docusaurus build process.',
        )
        .action(async () => {
          context.siteConfig?.plugins.forEach((pluginConfig) => {
            // Check PluginConfig is typed to [string, PluginOptions]
            if (pluginConfig && typeof pluginConfig[1] === 'object') {
              generateTypedoc(context, pluginConfig[1]);
            }
          });
        });
    },
  };
}

/**
 * Initiates a new typedoc Application bootstraped with plugin options
 */
async function generateTypedoc(context: any, opts: Partial<PluginOptions>) {
  const { siteDir } = context;

  const options = getPluginOptions(opts);

  const { id, docsRoot, ...optionsPassedToTypeDoc } = options;

  const outputDir = path.resolve(siteDir, options.docsRoot, options.out);

  const app = new Application();

  loadOptions(app);
  loadRenderer(app);
  loadTypedocPluginMarkdown(app);
  loadTypedocPluginFrontmatter(app);

  await app.bootstrapWithPlugins(
    optionsPassedToTypeDoc as unknown as Partial<TypeDocOptions>,
  );

  const project = app.convert();

  // if project is undefined typedoc has a problem - error logging will be supplied by typedoc.
  if (!project) {
    if (app.options.getValue('skipErrorChecking')) {
      return;
    }
    console.error(
      '[docusaurus-plugin-typedoc] TypeDoc exited with an error. Use the "skipErrorChecking" option to disable TypeDoc error checking.',
    );
    process.exit();
  }

  if (app.options.getValue('watch')) {
    app.convertAndWatch(async (project) => {
      await app.generateDocs(project, outputDir);
    });
  } else {
    await app.generateDocs(project, outputDir);
  }
}

export function writeFileSync(fileName: string, data: string) {
  fs.mkdirSync(path.dirname(normalizePath(fileName)), { recursive: true });
  fs.writeFileSync(normalizePath(fileName), data);
}

export function normalizePath(path: string) {
  return path.replace(/\\/g, '/');
}
