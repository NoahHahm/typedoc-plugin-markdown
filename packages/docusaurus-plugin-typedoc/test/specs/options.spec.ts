import { getPluginOptions } from '../../dist/options';
import { PluginOptions } from '../../src/models';

describe(`Options:`, () => {
  test(`should return default options`, () => {
    expect(
      getPluginOptions({ entryPoints: ['entrypoint.ts'] }),
    ).toMatchSnapshot();
  });
  test(`should merge plugins`, () => {
    expect(
      getPluginOptions({
        plugin: ['plugin-xyz'],
      }).plugin,
    ).toMatchSnapshot();
  });

  test(`should merge custom sidebar options`, () => {
    expect(
      getPluginOptions({
        sidebar: {
          position: 2,
        },
      } as PluginOptions).sidebar,
    ).toMatchSnapshot();
  });
});
