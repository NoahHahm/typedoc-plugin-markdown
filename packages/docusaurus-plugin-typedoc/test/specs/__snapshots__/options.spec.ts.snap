// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options: should merge custom sidebar options 1`] = `
{
  "autoConfiguration": true,
  "categoryLabel": "API",
  "collapsed": true,
  "fullNames": true,
  "indexLabel": "Custom index",
  "position": null,
}
`;

exports[`Options: should merge plugins 1`] = `
[
  "plugin-xyz",
]
`;

exports[`Options: should return default options 1`] = `
{
  "cleanOutputDir": true,
  "docsRoot": "docs",
  "enableEmojis": true,
  "enableFrontmatter": true,
  "entryDocument": "index.md",
  "entryPoints": [
    "entrypoint.ts",
  ],
  "hideBreadcrumbs": true,
  "hideInPageTOC": true,
  "hidePageTitle": false,
  "id": "default",
  "out": "api",
  "plugin": [
    "none",
  ],
  "sidebar": {
    "autoConfiguration": true,
    "categoryLabel": "API",
    "collapsed": true,
    "fullNames": false,
    "indexLabel": "Overview",
    "position": null,
  },
  "theme": "docusaurus",
  "watch": false,
}
`;
